### PR TITLE
Kompile temp dirs

### LINKED ISSUE #1059: Should different backends use different temporary directories?
I'm currently working on improving KTest's parallelism. One problem that I'm currently having is when I run `kompile definition.k` and `kompile definition.k --backend pdf`at the same time, they're having a race, whichever terminates first is removing the `.k` directory and other one is failing.

This is not a PDF backend specific problem though. If Java backend also uses temporary directory then `kompile definition.k` and `kompile definition.k --backend java` should also be a problem.

I'd like to discuss what to do about this. I think one easy way to fix this is to make backends generate temp directories with different names.

We briefly discussed about a filesystem API in #1008 . Maybe we can implement that too. Using the API we can make sure that backends won't be able to read from/write to arbitrary files, they'll be restricted with their temporary directory, which is generated by filesystem API by reading some fields of backend(to decide on a directory name). This will also help cleaning up after backends are done.

Any ideas?

### PR DESCRIPTION
**Pull request by osa1**
Resolves #1059 .

### CODE REVIEW

File: k-core/src/main/java/org/kframework/kompile/KompileFrontEnd.java [hunk 1]
```diff
@@ -104,12 +116,13 @@ public void run() {
...
 
         if (backend.generatesDefinition()) {
-                context.kompiled = new File(options.directory, FilenameUtils.removeExtension(options.mainDefinitionFile().getName()) + "-kompiled");
+            context.kompiled = new File(options.directory, FilenameUtils.removeExtension(
+                    options.mainDefinitionFile().getName()) + "-kompiled");
+            if (context.kompiled.exists()) {
                 checkAnotherKompiled(context.kompiled);
-                if (options.mainDefinitionFile().exists()
-                        && !context.kompiled.exists() && !context.kompiled.mkdirs()) {
-                    kem.registerCriticalError("Could not create directory " + context.kompiled);
```
Comment (osa1) [inline k-core/src/main/java/org/kframework/kompile/KompileFrontEnd.java:111]: This was printing "Could not create directory ..." when definition file is missing.

File: k-core/src/main/java/org/kframework/kompile/KompileFrontEnd.java [hunk 2]
```diff
@@ -104,12 +116,13 @@ public void run() {
         }
 
         if (backend.generatesDefinition()) {
-                context.kompiled = new File(options.directory, FilenameUtils.removeExtension(options.mainDefinitionFile().getName()) + "-kompiled");
+            context.kompiled = new File(options.directory, FilenameUtils.removeExtension(
+                    options.mainDefinitionFile().getName()) + "-kompiled");
+            if (context.kompiled.exists()) {
                 checkAnotherKompiled(context.kompiled);
```
Comment (osa1) [inline k-core/src/main/java/org/kframework/kompile/KompileFrontEnd.java:117]: This was checked even when parent directory is missing.

**Comment by osa1** [comment]
This is ready for reviews.

**Comment by radumereuta** [comment]
lgtm, but you should expect at least another review.

**Pull request merged**